cartex/src/runtime/mpi_comm/runtime.cpp

```diff
@@ -114,34 +138,331 @@ runtime::impl::neighborhood::neighborhood(
+unpack_y_kernel(runtime::real_type* field, runtime::real_type const* buffer_left,
+    runtime::real_type const* buffer_right, int halo_x_left, int halo_x_right, int halo_y_left,
+    int halo_y_right, int halo_z_left, int halo_z_right, int ext_x, int ext_y, int ext_z)
+{
+    const auto x = blockIdx.x * blockDim.x + threadIdx.x;
+    const auto z = blockIdx.z * blockDim.z + threadIdx.z;
+    const auto ext_x_b = ext_x + halo_x_left + halo_x_right;
+    if (x < ext_x_b && z < ext_z)
+    {
+        const auto ext_y_b = ext_y + halo_y_left + halo_y_right;
+        if (threadIdx.y < halo_y_left)
+        {
+            const auto y = threadIdx.y;
+            const auto offset = ext_x_b * (y + (z + halo_z_left) * ext_y_b);
+            const auto offset_l = ext_x_b * (y + halo_y_left * z);
+            field[x + offset] = buffer_left[x + offset_l];
+        }
+        else
+        {
+            const auto y = threadIdx.y - halo_y_left;
+            const auto offset = ext_x_b * (y + halo_y_left + ext_y + (z + halo_z_left) * ext_y_b);
+            const auto offset_r = ext_x_b * (y + halo_y_right * z);
+            field[x + offset] = buffer_right[x + offset_r];
+        }
+    }
+}
+#endif
+
+void
+runtime::impl::neighborhood::unpack_y(
+    memory_type& field, memory_type& buffer_left, memory_type& buffer_right)
+{
+#ifdef __CUDACC__
+    constexpr int block_dim_z = 4;
+    constexpr int block_dim_x = 64;
+    const dim3    block_dims(block_dim_x, m_halos[2] + m_halos[3], block_dim_z);
+    const dim3    num_blocks((ext_buffer[0] + block_dim_x - 1) / block_dim_x, 1,
+        (d.domain_ext[2] + block_dim_z - 1) / block_dim_z);
+    execute_kernel(num_blocks, block_dims, unpack_y_kernel, stream, field.hd_data(),
+        buffer_left.hd_data(), buffer_right.hd_data(), m_halos, d.domain_ext);
+#else
+    for (int z = 0; z < d.domain_ext[2]; ++z)
+    {
+        for (int y = 0; y < m_halos[2]; ++y)
+        {
+            const auto offset = ext_buffer[0] * (y + (z + m_halos[4]) * ext_buffer[1]);
+            const auto offset_l = ext_buffer[0] * (y + m_halos[2] * z);
+            for (int x = 0; x < ext_buffer[0]; ++x) field[x + offset] = buffer_left[x + offset_l];
+        }
+        for (int y = 0; y < m_halos[3]; ++y)
+        {
+            const auto offset = ext_buffer[0] * (y + m_halos[2] + d.domain_ext[1] +
+                                                    (z + m_halos[4]) * ext_buffer[1]);
+            const auto offset_r = ext_buffer[0] * (y + m_halos[3] * z);
+            for (int x = 0; x < ext_buffer[0]; ++x) field[x + offset] = buffer_right[x + offset_r];
+        }
+    }
+#endif
+}
+
+void
+runtime::impl::neighborhood::exchange(memory_type& field, std::vector<memory_type>& send_buffers,
+    std::vector<memory_type>& recv_buffers, int field_id)
+{
+    pack_x(field, send_buffers[0], send_buffers[1]);
+    CARTEX_CHECK_MPI_RESULT(
+        MPI_Sendrecv(send_buffers[1].hd_data(), send_buffers[1].size() * sizeof(runtime::real_type),
```

Comment:
Those two can be overlapped in `MPI_ISendrecv`, right? each two steps within the same dimension are independent, with a sync between the steps.

Reply:
In fact, the same applies to the MPI-types version.

Reply:
good point, will make it async

Reply:
... and then it will get slower ;P you never know!